frontend/src/__tests__/unit/useGroups.spec.ts

```diff
@@ -0,0 +1,123 @@
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(2);
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, true, true]);
+  });
+
+  it('should handle 403 as an empty result', async () => {
+    const error = {
+      statusObject: {
+        code: 403,
+      },
+    };
+    k8sListResourceMock.mockReturnValue(Promise.reject(error));
+
+    const renderResult = testHook(useGroups);
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], true, undefined, expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([false, false, true, true]);
+
+    // refresh
+    await act(() => renderResult.result.current[3]());
+    // error 403 should cache error and prevent subsequent attempts to fetch
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, true, true]);
+  });
+
+  it('should handle 404 as an error', async () => {
+    const error = {
+      statusObject: {
+        code: 404,
+      },
+    };
+    k8sListResourceMock.mockReturnValue(Promise.reject(error));
+
+    const renderResult = testHook(useGroups);
+
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expectHook(renderResult)
+      .toStrictEqual([[], false, new Error('No groups found.'), expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([true, true, false, true]);
+
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+
+    // refresh
+    await act(() => renderResult.result.current[3]());
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(2);
+    // we get a new error because the k8s API is called a 2nd time
+    expectHook(renderResult).toHaveUpdateCount(3).toBeStable([true, true, false, true]);
+  });
+
+  it('should handle other errors and rethrow', async () => {
+    const error = new Error();
+    k8sListResourceMock.mockReturnValue(Promise.reject(new Error()));
+
+    const renderResult = testHook(useGroups);
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, undefined, expect.any(Function)])
+      .toHaveUpdateCount(1);
+
+    // wait for update
+    await renderResult.waitForNextUpdate();
+    expect(k8sListResourceMock).toHaveBeenCalledTimes(1);
+    expectHook(renderResult)
+      .toStrictEqual([[], false, error, expect.any(Function)])
+      .toHaveUpdateCount(2)
+      .toBeStable([true, true, false, true]);
```

Comment:
```suggestion
      .toBeStable([false, true, false, true]);
```
Same here -- stability of the value should only happen on "re-renders" not invocations / calls finishing.

Reply:
For this test case, the result is the default value. Since we have an error in both updates, it continues to return the default value which is stable.

Reply:
Unless I am misremembering you fall into one of these categories:

1. You got an error, error changed, data stays the same
2. You got new data, data changed, no error
3. Exited early because we don't want to fetch
4. You're not ready for a change (failed dependencies of your check -- like `namespace` was not provided for fetching a namespaced resource)

I believe you fall into no3 with this test, no?

Reply:
At this line in particular, we've had 2 updates. The first is the request to fetch, and the second is when the error is thrown. So here we expect data to be stable and error to not be stable.

Afterwards a few lines below, a refresh call should result in a new error but again data will be the default value.

I'm making some changes in this area to assert return values include new errors thrown when we refresh.